File: mobile/src/main/java/com/windscribe/mobile/email/AddEmailActivity.java
/*
 * Copyright (c) 2021 Windscribe Limited.
 */

package com.windscribe.mobile.email;


import static com.windscribe.vpn.Windscribe.appContext;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Patterns;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.fragment.app.Fragment;
import androidx.work.Data;

import com.windscribe.mobile.R;
import com.windscribe.mobile.base.BaseActivity;
import com.windscribe.mobile.confirmemail.ConfirmActivity;
import com.windscribe.mobile.custom_view.ProgressFragment;
import com.windscribe.mobile.di.ActivityModule;
import com.windscribe.mobile.welcome.SoftInputAssist;
import com.windscribe.mobile.windscribe.WindscribeActivity;
import com.windscribe.vpn.commonutils.ThemeUtils;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.OnClick;

public class AddEmailActivity extends BaseActivity implements AddEmailView {

    @BindView(R.id.cl_add_email)
    ConstraintLayout mConstraintLayoutMain;

    @BindView(R.id.email_description)
    TextView mEmailDescription;

    @BindView(R.id.email)
    EditText mEmailEditView;

    @BindView(R.id.email_error)
    ImageView mEmailErrorView;

    @BindView(R.id.next)
    TextView mNextButton;

    @Inject
    AddEmailPresenter mPresenter;

    @BindView(R.id.nav_title)
    TextView mTitleView;

    private final String TAG = "[add_email_a]";

    private boolean goToHomeAfterFinish = false;

    private final TextWatcher mGeneralTextWatcher = new TextWatcher() {
        @Override
        public void afterTextChanged(Editable s) {
            if (s.length() > 0) {
                if (s.hashCode() == mEmailEditView.getText().hashCode()) {
                    mEmailDescription.setText(getString(R.string.email_description));
                    mEmailDescription.setTextColor(ThemeUtils.getColor(AddEmailActivity.this,R.attr.wdSecondaryColor,R.color.colorWhite50));
                    mEmailErrorView.setVisibility(View.GONE);
                    mEmailEditView.setTextColor(ThemeUtils.getColor(AddEmailActivity.this,R.attr.wdPrimaryColor,R.color.colorWhite50));
                }
            }
        }

        @Override
        public void beforeTextChanged(CharSequence s, int start, int count, int after) {

        }

        @Override
        public void onTextChanged(CharSequence s, int start, int before, int count) {
            if (start == 0 && before == 1 && count == 0) {
                mNextButton.setPressed(false);
            } else {
                mNextButton.setEnabled(
                        s.hashCode() == mEmailEditView.getText().hashCode() && Patterns.EMAIL_ADDRESS.matcher(s)
                                .matches());
            }
        }
    };

    private final Logger mMainLogger = LoggerFactory.getLogger(TAG);

    private SoftInputAssist softInputAssist;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setActivityModule(new ActivityModule(this, this)).inject(this);
        setContentLayout(R.layout.activity_add_email_address,true);

        goToHomeAfterFinish = getIntent().getBooleanExtra("goToHomeAfterFinish", false);
        mPresenter.setUpLayout();
    }

    @Override
    protected void onResume() {
        softInputAssist.onResume();
        super.onResume();
    }

    @Override
    protected void onPause() {
        softInputAssist.onPause();
        super.onPause();
    }

    @Override
    protected void onDestroy() {
        softInputAssist.onDestroy();
        mPresenter.onDestroy();
        super.onDestroy();
    }

    @Override
    public void gotoWindscribeActivity() {
        appContext.workManager.updateSession(Data.EMPTY);
        if (goToHomeAfterFinish) {
            Intent startIntent = WindscribeActivity.getStartIntent(this);
            startIntent.addFlags(Intent.FLAG_ACTIVITY_NO_HISTORY | Intent.FLAG_ACTIVITY_CLEAR_TOP);
            startActivity(startIntent);
        } else {
            startActivity(ConfirmActivity.getStartIntent(this));
            finish();
        }
    }

    @Override
    public void hideSoftKeyboard() {
        InputMethodManager imm = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
        if (imm != null) {
            imm.hideSoftInputFromWindow(getWindow().getDecorView().getWindowToken(),
                    InputMethodManager.HIDE_NOT_ALWAYS);
        }
    }

    @OnClick(R.id.next)
    public void onAddEmailClick() {
        mPresenter.onAddEmailClicked(mEmailEditView.getText() != null ? mEmailEditView.getText().toString() : null);
    }

    @OnClick(R.id.nav_button)
    public void onBackButtonPressed() {
        onBackPressed();
    }

    @Override
    public void prepareUiForApiCallFinished() {
        mMainLogger.info("Preparing ui for api call finished...");
        Fragment fragment = getSupportFragmentManager().findFragmentById(R.id.cl_add_email);
        if (fragment instanceof ProgressFragment) {
            ((ProgressFragment) fragment).finishProgress();
        }
    }

    @Override
    public void prepareUiForApiCallStart() {
        mMainLogger.info("Preparing ui for api call start...");
        ProgressFragment.getInstance().add(this, R.id.cl_add_email, true);
    }

    @Override
    public void setUpLayout(String title) {
        mEmailEditView.addTextChangedListener(mGeneralTextWatcher);
        softInputAssist = new SoftInputAssist(this, new int[]{});
        mTitleView.setText(title);
    }

    @Override
    public void showInputError(String errorText) {
        mEmailDescription.setTextColor(getResources().getColor(R.color.colorRed));
        mEmailDescription.setText(errorText);
        mEmailErrorView.setVisibility(View.VISIBLE);
        mEmailEditView.setTextColor(getResources().getColor(R.color.colorRed));
    }

    @Override
    public void showToast(String toastString) {
        Toast.makeText(this, toastString, Toast.LENGTH_SHORT).show();
    }
}

File: mobile/src/main/java/com/windscribe/mobile/email/AddEmailActivity.kt
/*
 * Copyright (c) 2021 Windscribe Limited.
 */
package com.windscribe.mobile.email

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.util.Patterns
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.EditText
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.work.Data
import butterknife.BindView
import butterknife.OnClick
import com.windscribe.mobile.R
import com.windscribe.mobile.base.BaseActivity
import com.windscribe.mobile.confirmemail.ConfirmActivity
import com.windscribe.mobile.custom_view.ProgressFragment
import com.windscribe.mobile.di.ActivityModule
import com.windscribe.mobile.welcome.SoftInputAssist
import com.windscribe.mobile.windscribe.WindscribeActivity.Companion.getStartIntent
import com.windscribe.vpn.Windscribe
import com.windscribe.vpn.commonutils.ThemeUtils.getColor
import org.slf4j.LoggerFactory
import javax.inject.Inject

class AddEmailActivity : BaseActivity(), AddEmailView {

    @BindView(R.id.cl_add_email)
    lateinit var constraintLayoutMain: ConstraintLayout

    @BindView(R.id.email_description)
    lateinit var emailDescription: TextView

    @BindView(R.id.email)
    lateinit var emailEditView: EditText

    @BindView(R.id.email_error)
    lateinit var emailErrorView: ImageView

    @BindView(R.id.next)
    lateinit var nextButton: TextView

    @BindView(R.id.nav_title)
    lateinit var titleView: TextView

    @Inject
    lateinit var presenter: AddEmailPresenter

    private val logger = LoggerFactory.getLogger("[add_email_a]")
    private var softInputAssist: SoftInputAssist? = null

    private val generalTextWatcher: TextWatcher = object : TextWatcher {
        override fun afterTextChanged(s: Editable) {
            if (s.isNotEmpty()) {
                if (s.hashCode() == emailEditView.text.hashCode()) {
                    emailDescription.text = getString(R.string.email_description)
                    emailDescription.setTextColor(
                        getColor(
                            this@AddEmailActivity,
                            R.attr.wdSecondaryColor,
                            R.color.colorWhite50
                        )
                    )
                    emailErrorView.visibility = View.GONE
                    emailEditView.setTextColor(
                        getColor(
                            this@AddEmailActivity,
                            R.attr.wdPrimaryColor,
                            R.color.colorWhite50
                        )
                    )
                }
            }
        }

        override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
        override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
            if (start == 0 && before == 1 && count == 0) {
                nextButton.isPressed = false
            } else {
                nextButton.isEnabled =
                    s.hashCode() == emailEditView.text.hashCode() && Patterns.EMAIL_ADDRESS.matcher(
                        s
                    )
                        .matches()
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setActivityModule(ActivityModule(this, this)).inject(this)
        setContentLayout(R.layout.activity_add_email_address, true)
        presenter.setUpLayout()
    }

    override fun onResume() {
        softInputAssist?.onResume()
        super.onResume()
    }

    override fun onPause() {
        softInputAssist?.onPause()
        super.onPause()
    }

    override fun onDestroy() {
        softInputAssist?.onDestroy()
        presenter.onDestroy()
        super.onDestroy()
    }

    override fun gotoWindscribeActivity() {
        Windscribe.appContext.workManager.updateSession(Data.EMPTY)
        if (intent.getBooleanExtra(finishAfterAddEmail, false)) {
            finish()
        } else if (intent.getBooleanExtra(goToHomeAfterFinish, false)) {
            val startIntent = getStartIntent(this)
            startIntent.addFlags(Intent.FLAG_ACTIVITY_NO_HISTORY or Intent.FLAG_ACTIVITY_CLEAR_TOP)
            startActivity(startIntent)
        } else {
            startActivity(ConfirmActivity.getStartIntent(this))
            finish()
        }
    }

    override fun hideSoftKeyboard() {
        val imm = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        imm.hideSoftInputFromWindow(
            window.decorView.windowToken,
            InputMethodManager.HIDE_NOT_ALWAYS
        )
    }

    @OnClick(R.id.next)
    fun onAddEmailClick() {
        if (emailEditView.text != null) {
            presenter.onAddEmailClicked(emailEditView.text.toString())
        }
    }

    @OnClick(R.id.nav_button)
    fun onBackButtonPressed() {
        onBackPressed()
    }

    override fun prepareUiForApiCallFinished() {
        logger.info("Preparing ui for api call finished...")
        val fragment = supportFragmentManager.findFragmentById(R.id.cl_add_email)
        if (fragment is ProgressFragment) {
            fragment.finishProgress()
        }
    }

    override fun prepareUiForApiCallStart() {
        logger.info("Preparing ui for api call start...")
        ProgressFragment.getInstance().add(this, R.id.cl_add_email, true)
    }

    override fun setUpLayout(title: String) {
        emailEditView.addTextChangedListener(generalTextWatcher)
        softInputAssist = SoftInputAssist(this, intArrayOf())
        titleView.text = title
    }

    override fun showInputError(errorText: String) {
        emailDescription.setTextColor(resources.getColor(R.color.colorRed))
        emailDescription.text = errorText
        emailErrorView.visibility = View.VISIBLE
        emailEditView.setTextColor(resources.getColor(R.color.colorRed))
    }

    override fun showToast(toastString: String) {
        Toast.makeText(this, toastString, Toast.LENGTH_SHORT).show()
    }

    companion object {
        const val finishAfterAddEmail = "finishAfterAddEmail"
        const val goToHomeAfterFinish = "goToHomeAfterFinish"
    }
}